AK/StringFloatingPointConversions.cpp

```diff
@@ -0,0 +1,2279 @@
+    { 287, 725, { 438889925503495095ull, 0ull } },
+    { 287, 726, { 877779851006990190ull, 0ull } },
+    { 287, 727, { 1755559702013980379ull, 0ull } },
+    { 287, 728, { 3511119404027960758ull, 0ull } },
+    { 288, 728, { 702223880805592152ull, 0ull } },
+    { 288, 729, { 1404447761611184303ull, 0ull } },
+    { 288, 730, { 2808895523222368606ull, 0ull } },
+    { 289, 730, { 561779104644473722ull, 0ull } },
+    { 289, 731, { 1123558209288947443ull, 0ull } },
+    { 289, 732, { 2247116418577894885ull, 0ull } },
+    { 290, 732, { 449423283715578977ull, 0ull } },
+    { 290, 733, { 898846567431157954ull, 0ull } },
+    { 290, 734, { 1797693134862315908ull, 0ull } },
+};
+
+// Checks if value is divisible by 5 ^ power.
+bool is_divisible_by_pow_5(u64 value, i32 power)
+{
+    static const u64 FIVE_POWERS[] = { 5, 25, 625, 390625, 152587890625 };
+
+    if (power <= 0 || value == 0) {
+        return true;
+    }
+    if (power >= 28) { // 2 ^ 64 - 1 < 5 ^ 28
+        return false;
+    }
+
+    i32 result = 0;
+    for (i32 i = 5; i--;) {
+        if (value % FIVE_POWERS[i] == 0) {
+            value /= FIVE_POWERS[i];
+            result += 1 << i;
+        }
+    }
+    return result >= power;
+}
+
+template<>
+FloatingPointExponentialForm convert_floating_point_to_fixed_decimal(double value)
+{
+    static const u64 LEN_M = 52;
+    static const i32 BIAS = 1023;
+    static const i32 LOOKUP_OFFSET = BIAS + LEN_M + 1;
+
+    u64 bit_representation = bit_cast<u64, double>(value);
+
+    bool sign = bit_representation >> 63;
+    i32 e = static_cast<i32>(bit_representation >> LEN_M) & ((1 << 11) - 1);
+    u64 m = bit_representation & ((1ull << LEN_M) - 1);
+
+    // For +0, it is {.sign = 0, fraction = 0, exponent = 0},
+    // for -0, is {.sign = 1, fraction = 0, exponent = 0},
+    if (e == 0 && m == 0) {
+        return { sign, 0, 0 };
+    }
+
+    // for +inf, -inf, and NaN is undefined.
+    // To be on the safe side, return something meaningless rather than just not checking.
+    if (e == 2047) [[unlikely]] {
+        return { 0, 0, 0 };
+    }
+
+    // Step 1. Decode the floating point number, and unify normalized and subnormal cases.
+    u64 mf = (e == 0 ? 0 : (1ull << LEN_M)) + m;
+    i32 ef = (e == 0 ? 1 : e) - BIAS - LEN_M;
+    // abs(value) = mf * 2 ^ ef
+
+    // Step 2. Determine the interval of information-preserving outputs.
+    i32 e2 = ef - 2; // -1076 <= e2 <= 969
+    u64 u = 4 * mf - (m == 0 && e > 1 ? 1 : 2);
+    u64 v = 4 * mf;
+    u64 w = 4 * mf + 2;
+    // u * 2 ^ e2 < abs(answer) < w * 2 ^ e2        (1)
+    // abs(value) = v * 2 ^ e2 (yet another representation)
+
+    // Computes `operand * multiplier >> shift_amount'
+    auto multiply_and_shift = [&](u64 operand, u64 multiplier[2], i32 shift) {
+        using u128 = unsigned __int128;
+
+        u128 a = static_cast<u128>(operand) * multiplier[0];
+        u128 b = static_cast<u128>(operand) * multiplier[1] + (a >> 64);
```

Comment:
```suggestion
        u128 b = static_cast<u128>(operand) * multiplier[1] + a.high();
```
maybe `a.low()` in the lines below instead of casting to `u64`, but both are equivalent